### PR TITLE
UefiCpuPkg/MmUnblockMemoryLib: Check if buffer range is valid

### PR DESCRIPTION
**Pull request by td36**
# Description
Check if input buffer range unblockable:
1.The input buffer range to block should be totally covered by one or multi memory allocation HOB
2.All the memory allocation HOB that overlap with the input buffer range should be EfiRuntimeServicesData, EfiACPIMemoryNVS or EfiReservedMemoryType.

## How This Was Tested

Tested in Intel internal server platform with the X86 standalone MM env and successfully booted into shell
Tested in QSP platform with the X86 standalone MM env and successfully booted into os

### CODE REVIEW

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c [hunk 1]
```diff
@@ -18,6 +18,84 @@
...
+
+  @param[in]  MemoryAllocationHob         Pointer to a memory allocation lib.
+  @param[in]  UnblockAddress              Base address of the range to unblock.
+  @param[in]  UnblockEnd                  End address of the range to unblock.
+
+  @retval RETURN_SUCCESS              The range to unblock is valid.
+  @retval RETURN_INVALID_PARAMETER    The range to unblock contains invalid memory.
+**/
+EFI_STATUS
+CheckIsUnblockRangeValid (
```
Comment (niruiyu) [inline UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c:36]: MmUnblockMemoryLibIsUnblockableRegion()

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c [hunk 2]
```diff
@@ -18,6 +18,84 @@
...
+  EFI_STATUS            Status;
+
+  while (MemoryAllocationHob != NULL) {
+    HobBase = MemoryAllocationHob->AllocDescriptor.MemoryBaseAddress;
+    HobEnd  = MemoryAllocationHob->AllocDescriptor.MemoryBaseAddress +
+              MemoryAllocationHob->AllocDescriptor.MemoryLength;
+    if ((UnblockBase < HobEnd) && (UnblockEnd > HobBase)) {
+      if (!((MemoryAllocationHob->AllocDescriptor.MemoryType == EfiRuntimeServicesData) ||
+            (MemoryAllocationHob->AllocDescriptor.MemoryType == EfiACPIMemoryNVS) ||
+            (MemoryAllocationHob->AllocDescriptor.MemoryType == EfiReservedMemoryType)))
```
Comment (niruiyu) [inline UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c:53]: Type != RT && Type != NVS && Type != Reserved

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c [hunk 3]
```diff
@@ -45,12 +127,20 @@ MmUnblockMemoryRequest (
...
   EFI_PEI_MM_COMMUNICATION_PPI  *MmCommunicationPpi;
+  EFI_PEI_HOB_POINTERS          Hob;
 
   if (!IS_ALIGNED (UnblockAddress, SIZE_4KB)) {
     DEBUG ((DEBUG_ERROR, "Error: UnblockAddress is not 4KB aligned: %p\n", UnblockAddress));
     return EFI_INVALID_PARAMETER;
   }
 
+  Hob.Raw = GetFirstHob (EFI_HOB_TYPE_MEMORY_ALLOCATION);
+  ASSERT (Hob.Raw != NULL);
```
Comment (niruiyu) [inline UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c:138]: Move the GetFirstHob() call inside the sub-function.

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c [hunk 4]
```diff
@@ -18,6 +18,84 @@
...
+        DEBUG ((DEBUG_ERROR, "Error: range [0x%lx, 0x%lx] to unblock contains invalid type memory\n", UnblockBase, UnblockEnd));
+        return RETURN_INVALID_PARAMETER;
+      }
+
+      if (UnblockBase < HobBase) {
+        //
+        // Recursively call to check [UnblockBase, HobBase]
+        //
+        Status = CheckIsUnblockRangeValid (
+                   (EFI_HOB_MEMORY_ALLOCATION *)GetNextHob (EFI_HOB_TYPE_MEMORY_ALLOCATION, GET_NEXT_HOB (MemoryAllocationHob)),
```
Comment (niruiyu) [inline UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c:67]: Just pass in GET_NEXT_HOB ()

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c (hunk 1 above)
Comment (td36) [reply UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c]: Changed to MmUnblockMemoryLibIsUnblockableRegion

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c (hunk 2 above)
Comment (td36) [reply UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c]: Changed here

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c (hunk 3 above)
Comment (td36) [reply UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c]: GetHobList () is passed into sub-function here

File: UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c (hunk 4 above)
Comment (td36) [reply UefiCpuPkg/Library/MmUnblockMemoryLib/MmUnblockMemoryLib.c]: Changed here.

**Review by niruiyu — APPROVED** [review]

**Pull request merged**